### PR TITLE
Provide a PHP Api for UserStatus

### PR DESCRIPTION
**Pull request by georgehrke**
This PR provides a PHP API for the new User Status feature.
It allows apps to query the status for a batch of users.

### CODE REVIEW

**Comment by georgehrke** [comment]
@nickvergessen Does this look reasonable for your use-case in Talk?

File: apps/user_status/lib/Connector/UserStatus.php [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+use OCP\IUser;
+use OCP\UserStatus\IUserStatus;
+use OCA\UserStatus\Db;
+
+class UserStatus implements IUserStatus {
+
+	/** @var IUser */
+	private $user;
+
+	/** @var string|null */
```
Comment (georgehrke) [inline apps/user_status/lib/Connector/UserStatus.php:37]: Should be string

File: apps/user_status/lib/Connector/UserStatusProvider.php [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+		foreach ($users as $user) {
+			$userId = $user->getUID();
+			$userMap[$userId] = $user;
+			$userIds[] = $userId;
+		}
+
+		return array_map(static function ($status) use ($userMap) {
+			$userId = $status->getUserId();
+			return new UserStatus($userMap[$userId], $status);
+		}, $this->service->findByUserIds($userIds));
```
Comment (georgehrke) [inline apps/user_status/lib/Connector/UserStatusProvider.php:59]: Maybe rather return an associative array with userId => IUserStatus here?

File: apps/user_status/lib/Connector/UserStatus.php [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+	public function __construct(IUser $user,
+								Db\UserStatus $status) {
+		$this->user = $user;
+		$this->status = $status->getStatus();
+		$this->message = $status->getCustomMessage();
+		$this->icon = $status->getCustomIcon();
+
+		if ($status->getClearAt() !== null) {
+			$this->clearAt = DateTimeImmutable::createFromFormat('U', (string)$status->getClearAt());
+		}
```
Comment (georgehrke) [inline apps/user_status/lib/Connector/UserStatus.php:64]: ToDo: automatically replace invisible with offline here.

Comment (georgehrke) [reply apps/user_status/lib/Connector/UserStatus.php:67]: Done

**Comment by georgehrke** [comment]
upon the request of @nickvergessen, i changed it from `IUser` to `userId`

File: lib/public/UserStatus/IManager.php [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+	 * @param string[] $userIds
+	 * @return IUserStatus[]
+	 * @since 20.0.0
+	 */
+	public function getUserStatuses(array $userIds):array;
+
+	/**
+	 * @param string $class
+	 * @since 20.0.0
+	 * @internal
```
Comment (nickvergessen) [inline lib/public/UserStatus/IManager.php:48]: If it's internal don't add it to the public API. you can still use it later on.

File: lib/public/UserStatus/IUserStatus.php [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+	 *
+	 * @return string|null
+	 * @since 20.0.0
+	 */
+	public function getIcon():?string;
+
+	/**
+	 * Gets the time that the custom status will be cleared at
+	 *
+	 * @return DateTimeImmutable
```
Comment (nickvergessen) [inline lib/public/UserStatus/IUserStatus.php:73]:
```suggestion
	 * @return DateTimeImmutable|null
```

**Review by nickvergessen — APPROVED** [review]
:shark:

**Review by MorrisJobke — APPROVED** [review]
Code looks good 👍

**Pull request merged**